### PR TITLE
Add filtering expressions to Dedalus rules. #178

### PR DESCRIPTION
**Pull request by tylerhou**
We only support expressions between identifiers currently. In the future we will add literal support and arithmetic expression support to both the LHS and RHS of the boolean expression.

We use | to denote the start of the filtering operator because using a comma introduces ambiguity with starting another Atom.

Filtering expressions also currently do not affect codegen.

### CODE REVIEW

File: hydroflow_datalog/src/grammar.rs [hunk 1]
```diff
@@ -17,28 +17,36 @@ pub mod datalog {
...
         _from: (),
+
         #[rust_sitter::repeat(non_empty = true)]
         #[rust_sitter::delimited(
             #[rust_sitter::leaf(text = ",")]
             ()
         )]
         pub sources: Vec<Atom>,
+
+        pub filters: Option<RuleFilters>,
```
Comment (shadaj) [inline hydroflow_datalog/src/grammar.rs:31]: This requires all filters to come after sources, but it'd be nice to instead make `Atom` an enum that has variants for both reading from relations or boolean conditions. That would also resolve the issue with needing different delimiters.

File: hydroflow_datalog/src/lib.rs [hunk 2]
```diff
@@ -352,4 +352,16 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_simple_filter() {
+        test_datalog_snapshot!(
+            r#"
+            .input input
+            .output out
+
+            out(x, y) :- input(x, y) | { x > y }.
```
Comment (shadaj) [inline hydroflow_datalog/src/lib.rs:363]: Once you add the code-gen logic for boolean expressions, we should also add runtime tests. Let's try to get those into this PR.

**Review by shadaj — CHANGES_REQUESTED** [review]
This is a great start! A little feedback on how we structure the AST and we should get the codegen logic into this PR.

**Comment by tylerhou** [comment]
Runtime tests added!.

File: hydroflow_datalog/src/grammar.rs [hunk 3]
```diff
@@ -54,4 +66,26 @@ pub mod datalog {
...
+    pub enum BoolOp {
+        Lt(#[rust_sitter::leaf(text = "<")] ()),
+        LtEq(#[rust_sitter::leaf(text = "<=")] ()),
+        Gt(#[rust_sitter::leaf(text = ">")] ()),
+        GtEq(#[rust_sitter::leaf(text = ">=")] ()),
+        Eq(#[rust_sitter::leaf(text = "==")] ()),
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct PredicateExpr {
```
Comment (shadaj) [inline hydroflow_datalog/src/grammar.rs:80]: Part of me wonders if we should also support arbitrary Rust expressions, maybe using backticks as delimiters. So  you could write expressions like:

```
out(x, y) :- in(x, y), (`my_rust_function(x)` > y)
```

Thoughts? We could leave this to a follow-up PR.

**Review by shadaj — CHANGES_REQUESTED** [review]
Looks great overall!

**Comment by jhellerstein** [comment]
Can we close this issue? @shadaj

**Comment by shadaj** [comment]
This is basically ready to merge minus a couple of test cleanups, so let's keep this open.

File: hydroflow_datalog/src/grammar.rs (hunk 3 above)
Comment (tylerhou) [reply hydroflow_datalog/src/grammar.rs:80]: Follow up PR. Like to merge something sooner than later at least... this PR has taken like four months so far now :).

File: hydroflow/tests/datalog_frontend.rs [hunk 4]
```diff
@@ -245,3 +245,144 @@ pub async fn test_local_constraints() {
...
+#[tokio::test]
+pub async fn test_boolean_relation_ge() {
+    let (in_send, input) = hydroflow::util::unbounded_channel::<(usize, usize)>();
+    let (out, mut out_recv) = hydroflow::util::unbounded_channel::<(usize, usize)>();
+
+    in_send.send((1, 1)).unwrap();
+    in_send.send((1, 2)).unwrap();
+    in_send.send((2, 1)).unwrap();
+
+    thread::spawn(|| {
```
[the remark this answers is not in the document]
Comment (tylerhou) [reply hydroflow/tests/datalog_frontend.rs:370]: I'll do this soon.

[the remark this answers is not in the document]
Comment (tylerhou) [reply hydroflow/tests/datalog_frontend.rs:370]: Done.

**Review by shadaj — APPROVED** [review]

**Pull request merged**